lib/srv/app/gcp/handler.go

```diff
@@ -0,0 +1,276 @@
+		// header to be the same as the outbound request's URL host.
+		forward.PassHostHeader(false),
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	svc.fwd = fwd
+	return svc, nil
+}
+
+// RoundTrip handles incoming requests and forwards them to the proper API.
+func (s *handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	if err := s.serveHTTP(w, req); err != nil {
+		s.formatForwardResponseError(w, req, err)
+		return
+	}
+}
+
+// serveHTTP is a helper to simplify error handling in ServeHTTP.
+func (s *handler) serveHTTP(w http.ResponseWriter, req *http.Request) error {
+	sessionCtx, err := common.GetSessionContext(req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	fwdRequest, err := s.prepareForwardRequest(req, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	recorder := httplib.NewResponseStatusRecorder(w)
+	s.fwd.ServeHTTP(recorder, fwdRequest)
+	status := uint32(recorder.Status())
+
+	if err := sessionCtx.Audit.OnRequest(req.Context(), sessionCtx, fwdRequest, status, nil); err != nil {
+		// log but don't return the error, because we already handed off request/response handling to the oxy forwarder.
+		s.Log.WithError(err).Warn("Failed to emit audit event.")
+	}
+	return nil
+}
+
+func (s *handler) formatForwardResponseError(rw http.ResponseWriter, r *http.Request, err error) {
+	s.Log.WithError(err).Debugf("Failed to process request.")
+	common.SetTeleportAPIErrorHeader(rw, err)
+
+	// Convert trace error type to HTTP and write response.
+	code := trace.ErrorToCode(err)
+	http.Error(rw, http.StatusText(code), code)
+}
+
+// prepareForwardRequest prepares a request for forwarding, updating headers and target host. Several checks are made along the way.
+func (s *handler) prepareForwardRequest(r *http.Request, sessionCtx *common.SessionContext) (*http.Request, error) {
+	forwardedHost := r.Header.Get("X-Forwarded-Host")
+	if !gcp.IsGCPEndpoint(forwardedHost) {
+		return nil, trace.AccessDenied("%q is not a GCP endpoint", forwardedHost)
+	}
+
+	payload, err := awsutils.GetAndReplaceReqBody(r)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	reqCopy, err := http.NewRequest(r.Method, r.URL.String(), bytes.NewReader(payload))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	reqCopy.URL.Scheme = "https"
+	reqCopy.URL.Host = forwardedHost
+	reqCopy.Header = r.Header.Clone()
+
+	err = s.replaceAuthHeaders(r, sessionCtx, reqCopy)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return reqCopy, trace.Wrap(err)
+}
+
+func (s *handler) replaceAuthHeaders(r *http.Request, sessionCtx *common.SessionContext, reqCopy *http.Request) error {
+	auth := reqCopy.Header.Get("Authorization")
+	if auth == "" {
+		s.Log.Debugf("No Authorization header present, skipping replacement.")
+		return nil
+	}
+
+	s.Log.Debugf("Processing request, sessionId = %q, gcpServiceAccount = %q", sessionCtx.Identity.RouteToApp.SessionID, sessionCtx.Identity.RouteToApp.GCPServiceAccount)
```

Comment:
Why this log function is in `replaceAuthHeaders` ? 
this function logs  about general request processing but the scope of this function is related to auth header replacement.

Reply:
Agreed, I'll move it.

Reply:
Done.